src/Microsoft.Extensions.Options/IOptionsCache.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.Options
+{
+    /// <summary>
+    /// Used to cache TOptions instances.
+    /// </summary>
+    /// <typeparam name="TOptions">The type of options being requested.</typeparam>
+    public interface IOptionsCache<TOptions> where TOptions : class
+    {
+        TOptions GetOrAdd(string name, Func<TOptions> createOptions);
+
+        bool TryAdd(string name, TOptions options);
+
+        bool TryRemove(string name);
+
+        // Do we need a Clear all?
```

Comment:
At some point it might be required. 

I also was thinking about using some kind of `IEvictionPolicy` and `TryAdd(TOptions, IEvictionPolicy<TOptions>)`. In that case we might add `ClearEvicted` method (perhaps we no need it at all (or no need right now), but just like a thing to consider).